### ISSUE TITLE
Killed by kernel due to memory exhaustion (32GB Memory + 32GB Swap)

### ISSUE THREAD

**Issue by frankyan** [gpertea/stringtie#9]
I am using stringtie 1.0.2 on a ubuntu server, which has 32GB memory and 32GB swap, 8 core cpu. When I am processing a bam file generated by HISAT, the stringtie process always be killed by linux kernel due to 32GB memory exhaustion. I checked the log file and found it it always stoped here:
```[04/04 22:09:10]>bundle chr7:46443159-46919933(2081294) (2027js, 19 guides) loaded, begins processing...``` 
In this bundle, mem and swap usage always increased to 100%. It seems that 2081294 means reads number, I guess the `-s` option may affect the memory usage. So I change `-s` default number to `-s 500000`. But it didn't work.  
Can you help me to figure out what caused the problem and how to fix it?

**Comment by gpertea** [comment]
It does seem that you have a very large bundle there, with a lot of reads and junctions. It may be that some dubious HISAT alignments joined multiple highly expressed genes in that region. We did notice before that sometimes HISAT alignments may need more aggressive filtering in order to prevent the formation of over-sized bundles like these.

I'd like to take a closer look at that particular bundle and implement better (user controlled) filtering for cases like this, when memory usage can be an issue. Is it possible to share that part of the .bam file and the related reference transcripts (is it a common human annotation)? 

If you agree, please write to me directly at gpertea@jhu.edu and I will provide instructions for extracting that region of the .bam and uploading that .bam region to our ftp server.
Thanks!

**Comment by frankyan** [comment]
Thanks for your response. I checked the bam file with IGV, found 2 possible causes. 
First, in this region there are huge amounts of reads which are non-uniquely mapped. Should I use StringTie with an extracted uniq-mapped bam file?  
Second, there are too many unbelievable long range junctions, such as CIGAR=32M70275N94M. This is my data's fault, the other 14 data didn't show such long junctions.
I can extract a part of bam file. I will contact you via email.

**Comment by frankyan** [comment]
It's mouse RNA data with mm10 reference(sequence and annotation)

**Comment by gpertea** [comment]
Thanks for sending your the BAM alignments in that region to us. Indeed, some of the alignments we see there have weird spliced alignments around polyA/polyT terminal stretches, linking relatively distant genes (albeit highly expressed) in the region . You might be able to get rid of those bad alignments if you first trim the polyAs using something like fqtrim, before running HISAT. Running on an uniq-mapped BAM file might help too, but may not fix the "hard-to-believe introns" issue which falsely connect relatively distant genes (this may be something for @infphilo to take a look at?).

However running StringTie on that particular BAM file on our machines only amounted to a memory use up to around 6GB. So I suspect there might be other large bundle(s) nearby which probably compounded the memory problem on your server, ultimately causing the memory exhaustion - especially if you ran StringTie with the -p option to have multiple threads process multiple (large) bundles at the same time? (this should be visible in the stderr messages with -v). Even without -p, StringTie still uses an additional background thread to read ahead from the .BAM file and build the next bundle and prepare it for processing, while the current bundle is being processed.. So if the next bundle is (much) larger than the one you sent us, the memory use can increase some more, even though the stderr messages only show the current bundle being processed.

So if you used -p when it crashed, perhaps you could give it another try without -p, with the latest version (v1.0.3 - as we fixed some potentially dangerous memory issues, though in our tests on your data it was about the same memory use with both v1.0.2 and v1.0.3). Or, if you prefer, we'd be glad to take an even closer look at the possible issues with your data so it would be great if you could actually send us the whole .BAM file you used -- or at least the whole chr17 and chr18 mappings (assuming that in the BAM file chr18 alignments actually follow chr17).

**Comment by frankyan** [comment]
Thanks for your advice. I found that the collaborator  gave me the raw data with adapters and polyA contamination. I cleaned the data and mapped with hisat and tophat. Both got lots of weird spliced alignments. But this time, stringtie worked well and fast on the tophat generated BAM file.  However, for hisat's BAM file, stringtie still ate lots of memory and took too long time to run , so I killed it. And it's a little strange in the stringtie logfile. After processing chrY, it went back to chr12, is it normal? Then it took very long time to process chr7, I killed the process finally.  
```
[04/11 21:01:47]>bundle chrY:90724455-90822305(683) (52js, 1 guides) loaded, begins processing...
24325 [04/11 21:01:47]^bundle chrY:90724455-90822305(683) done (5 processed potential transcripts).
24326 [04/11 21:01:51] 17699263 aligned fragments found.                              
24327 [04/11 21:02:22]^bundle chr12:103532565-104418579(603600) done (156 processed potential transcripts).
24328 [04/11 21:03:28]^bundle chr19:4029846-4201603(525184) done (83 processed potential transcripts).
24329 [04/11 21:22:27]^bundle chr7:46644641-46857665(2150788) done (74 processed potential transcripts).
```
I think data quality and HISAT are main causes.  If you still need my data for testing, I can upload the whole bam file or even the fastq file to your ftp.

**Comment by gpertea** [comment]
HISAT is still being refined and even though its speed and accuracy are great there are still some problems with some of the generated alignments being more "noisy" than TopHat. So while we are still trying to find better ways to filter some of these improper alignments from HISAT, I would advise trying to generate the alignments with TopHat, if possible (though I know it's much slower).

Now it seems clear that you used the multi-CPU option (-p) when you ran StringTie on the HISAT data, which explains also the higher memory usage that you ran into, compared to what we observed on your data (since for debugging purposes we ran it without the -p option). So if you run out of memory the first thing to try is to remove the -p option and you might be able to finish the run (though obviously it'll take longer).

The order of the bundles you see listed there is actually the order in which the various threads **finished** ("done") processing their heavy bundles. This is normal for multi-threaded runs of StringTie, especially with large noisy bundles like these - various bundles will take various times to run (some a lot longer than others).
